packages/smooth_app/lib/pages/onboarding/onboarding_flow_navigator.dart

```diff
@@ -0,0 +1,84 @@
+import 'package:flutter/cupertino.dart';
+import 'package:flutter/material.dart';
+import 'package:smooth_app/data_models/user_preferences.dart';
+import 'package:smooth_app/pages/onboarding/scan_example.dart';
+import 'package:smooth_app/pages/onboarding/welcome_page.dart';
+import 'package:smooth_app/pages/page_manager.dart';
+
+enum OnboardingPage {
+  NOT_STARTED,
+  WELCOME,
+  SCAN_EXAMPLE,
+  ONBOARDING_COMPLETE,
+}
+
+/// Decide which page to take the user to.
+class OnboardingFlowNavigator {
+  OnboardingFlowNavigator(this._userPreferences);
+
+  final Future<UserPreferences> _userPreferences;
+
+  Widget getNextPageWidget(BuildContext context, OnboardingPage currentPage) {
+    switch (currentPage) {
+      case OnboardingPage.NOT_STARTED:
+        // First screen, doesn't have a back navigation button.
+        return const WelcomePage();
+        break;
+      case OnboardingPage.WELCOME:
+        return _wrapWidgetInCustomBackNavigator(
+          context,
+          currentPage,
+          const ScanExample(),
+        );
+        break;
+      case OnboardingPage.SCAN_EXAMPLE:
+      case OnboardingPage.ONBOARDING_COMPLETE:
+        return PageManager();
+    }
+  }
+
+  void navigateToNextPage(BuildContext context, OnboardingPage currentPage) {
+    _userPreferences.then((UserPreferences prefs) {
+      prefs.setLastVisitedOnboardingPage(currentPage);
+      Navigator.push<Widget>(
+        context,
+        MaterialPageRoute<Widget>(
+          builder: (BuildContext context) =>
+              getNextPageWidget(context, currentPage),
+        ),
+      );
+    });
+  }
+
+  OnboardingPage _getPrevPage(OnboardingPage currentPage) {
+    switch (currentPage) {
+      case OnboardingPage.NOT_STARTED:
+      case OnboardingPage.WELCOME:
+        return OnboardingPage.NOT_STARTED;
+      case OnboardingPage.SCAN_EXAMPLE:
+        return OnboardingPage.WELCOME;
+      case OnboardingPage.ONBOARDING_COMPLETE:
+        return OnboardingPage.SCAN_EXAMPLE;
+    }
+  }
+
+  Widget _wrapWidgetInCustomBackNavigator(
+      BuildContext context, OnboardingPage currentPage, Widget widget) {
+    return WillPopScope(
```

Comment:
I'm not sure I've understood everything, but it looks like 1. there's no "Previous" button and 2. you override the "back" button in order to emulate the "Previous" button.

If so, I'm not a big fan:
1. it would be much more visible for the end-user to see an explicit "Previous" button
2. if for some reason I'm in stage 2 of onboarding, then I go to another app, and then I go back to smoothie, OK stage 2 of onboarding will be displayed, but if I click on the back button I would expect to exit the app, not to go back to stage 1.
3. I haven't checked if there's no infinite loop when you're on stage 1 and click on the back button (because the previous stage of stage 1 is stage 1)

Reply:
I was going with the back button that all flutter apps have in the app bar. We also have this in our app on multiple screens so it's consistent. But good point still, I will have a chat with the UX designer and see what he thinks

Re:
1. and 2. maybe you are right, but tbh I'm not sure about it, might need a user study, I'll send an email to the UXer and send a fix if he agrees with you.
3. there's no back button on the first onboarding page.

![Screenshot_1640606127](https://user-images.githubusercontent.com/7541730/147475009-b43648d0-5d07-4b74-b272-f9560b83bbb6.png)

Reply:
sent a mail to them, let me know your email account if you would like me to include you

Reply:
I was mentioning not just the left arrow button in the app bar, but the standard back "action" (e.g. on Android with one on the 3 buttons in the bottom bar).

Reply:
oh I see, hmm that functioning we can't override, so it will do whatever is appropriate (like closing the app or going back). I can test it in a bit (right now not on computer)